### ISSUE TITLE
Changing cron expression executes the job automatically even the execution time is too far in the future.

### ISSUE THREAD

**Issue by ahmedanwar100** [HangfireIO/Hangfire#2371]
### Background:
We have a project named **ScheduleJobs** that contains all the recurring jobs and we fetch the cron expressions from another API on application startup.
When the cron expression is changed and the **ScheduleJobs** project is restarted through **IIS**, the job (whose expression is changed) runs immediately even the actual execution time is too far in the future.

**Note:**
We are using SQL Server for job state storage.

### Actual Behaviour:
Hangfire should executes the job on its actual time.

**Comment by odinserj** [comment]
Hi, what version of Hangfire.Core you are using, and can you show an example of `AddOrUpdate` method call you are using?

**Comment by ahmedanwar100** [comment]
Hangfire.Core: 1.7.31
```
 RecurringJob.AddOrUpdate<ServiceA>(servce =>
                   servce.SetUsersInactive(),
                       cronExpressions.FirstOrDefault(cron=>cron.JobName=="UsersInactiveJob"),
                   TimeZoneInfo.Local,
                   HangfireConstants.CustomQueue1);
```

**Comment by odinserj** [comment]
Thank you for the details. Unfortunately I'm unable to replicate this behavior. Can you please create a repository on GitHub that shows the problem?

**Comment by ahmedanwar100** [comment]
Unfortunately, I was also not able to produce it again but that bug came up on a customer env. when the **ScheduleJobs** application hosted on IIS is restarted.

**Comment by odinserj** [comment]
I was finally able to reproduce this behavior and commit fixes in 1.8.X. Now thinking about backporting these changes to 1.7.X branch, however this will likely happen next week only.

**Comment by ahmedanwar100** [comment]
How did you reproduce this? And what was the fix? Can you please explain?

**Comment by odinserj** [comment]
This is a two step process, the problem doesn't occur directly after a first call to `AddOrUpdate`, but requires the following calls:

1. `AddOrUpdate` with a new cron expression that causes recurring job to have `NextExecution` pointing to some distant future, and `LastExecution` from the previous state where job scheduled with a previous cron expression.
2. Simple `AddOrUpdate` call that doesn't change anything. If a new cron expression has occurrence between `LastExecution` (from previous cron expression) and `NextExecution` that was changed for a new cron expression, its `NextExecution` will be updated anyway in previous versions, and not on the now datetime, but on the `LastExecution`'s value.

I refactored recurring job implementation to have less side effects, and now it updates `NextExecution` only when cron expression or time zone is changed, and it doesn't allow recurring job to be re-scheduled to the past in this case. so (1) step works as expected, and (2) step doesn't cause schedule changes.
